ledger/store/store.go

```diff
@@ -81,44 +79,67 @@ func NewStore(log zerolog.Logger, opts ...Option) (*Store, error) {
 		db:  db,
 		tx:  db.NewTransaction(true),
 
-		sema:  semaphore.NewWeighted(16),
-		err:   make(chan error, 16),
-		done:  make(chan struct{}),
-		mutex: &sync.RWMutex{},
-		wg:    &sync.WaitGroup{},
+		sema: semaphore.NewWeighted(16),
+		err:  make(chan error, 16),
+		done: make(chan struct{}),
+		txMu: &sync.RWMutex{},
+		wg:   &sync.WaitGroup{},
+
+		isDirtyMu: &sync.RWMutex{},
+		isDirty:   make(map[hash.Hash]bool, config.CacheSize),
 	}
 
 	s.wg.Add(1)
 	go s.flush()
 
-	s.cache, err = lru.NewWithEvict(config.CacheSize, func(k interface{}, v interface{}) {
+	s.cache, err = lru.NewWithEvict(config.CacheSize, func(k, v interface{}, used int) bool {
 		hash, ok := k.(hash.Hash)
 		if !ok {
 			logger.Fatal().Interface("got", k).Msg("unexpected key format")
 		}
 
-		value, ok := v.(ledger.Value)
-		if !ok {
-			logger.Fatal().Interface("got", v).Msg("unexpected value format")
+		// If the entry is dirty and the cache is not full, abort eviction.
+		s.isDirtyMu.RLock()
+		if s.isDirty[hash] && used < s.cacheSize {
+			s.isDirtyMu.RUnlock()
+			return false
 		}
+		s.isDirtyMu.RUnlock()
 
-		err := s.write(hash, value)
-		if err != nil {
-			logger.Fatal().Err(err).Msg("could not persist ledger payload")
+		// If the cache is full of dirty entries, push the current commit to disk and evict the entry.
+		if s.cacheFullAndDirty(used) {
+			s.forceCommit()
 		}
+
+		// If the current entry is clean, we can evict it. Otherwise, abort the eviction.
+		s.isDirtyMu.RLock()
+		defer s.isDirtyMu.RUnlock()
+		return !s.isDirty[hash]
 	})
 	if err != nil {
 		return nil, fmt.Errorf("could not create cache storage for ledger payloads: %w", err)
 	}
 
-	go s.persist()
-
 	return &s, nil
 }
 
 // Save stores a payload.
-func (s *Store) Save(hash hash.Hash, payload *ledger.Payload) {
+func (s *Store) Save(hash hash.Hash, payload *ledger.Payload) error {
+	// Store in cache.
 	_ = s.cache.Add(hash, payload.Value)
+
+	// Set state to dirty.
+	s.isDirtyMu.Lock()
+	s.isDirty[hash] = true
+	s.isDirtyMu.Unlock()
```

Comment:
We should do that _after_ writing it in the transaction. Otherwise we have a dirty value that can never become clean, right? Better to guard the entire method with the mutex then.